pilot/pkg/proxy/envoy/v2/eds.go

```diff
@@ -125,14 +220,222 @@ func localityLbEndpointsFromInstances(instances []*model.ServiceInstance) []endp
+			if node == "" && discReq.Node != nil {
+				node = discReq.Node.Id
+			}
+
+			if node == "" {
+				log.Infoa("Missing node ", discReq.String())
+				return errors.New("Missing node " + discReq.String())
+			}
+
+			// This doesn't seem to happen - add a log just in case. Typical requests from envoy
+			// have a single cluster, and
+			if len(con.Clusters) > 0 && len(clusters2) > 0 && con.Clusters[0] != clusters2[0] {
+				log.Infof("EDS: Clusters override %v -> %v %s", con.Clusters, clusters2, discReq.String())
+			}
+			// Doesn't seem to happen - logging in case something changes in envoy
+			if len(clusters2) > 1 {
+				log.Infof("EDS: Batch clusters %v %s", clusters2, discReq.String())
+			}
+			// Given that Pilot holds an eventually consistent data model, Pilot ignores any acknowledgements
+			// from Envoy, whether they indicate ack success or ack failure of Pilot's previous responses.
+			// Only the first request is actually horored and processed. Pilot drains all other requests from this stream.
+			if discReq.ResponseNonce != "" {
+				// TODO: once the deps are updated, log the ErrorCode if set (missing in current version)
+				if edsDebug {
+					log.Infof("EDS: ACK %s %s %s", node, discReq.VersionInfo, con.Clusters)
+				}
+				continue
+			} else {
+				if len(con.Clusters) > 0 {
+					// Should not happen
+					log.Infof("EDS REQ repeated %s %v/%v %v raw: %s ",
+						node, clusters2, con.Clusters, peerAddr, discReq.String())
+				}
+				// Initial request
+				if edsDebug {
+					log.Infof("EDS REQ %s %v %v raw: %s ",
+						node, clusters2, peerAddr, discReq.String())
+				}
+			}
+			if len(clusters2) > 0 {
+				con.Clusters = clusters2
+			}
+
+			for _, c := range con.Clusters {
+				s.addEdsCon(c, node, con)
+			}
+
+		case <-con.pushChannel:
+			if len(con.Clusters) == 0 {
+				continue
+			}
+		}
+
+		if len(con.Clusters) > 0 {
+			response := s.endpoints(s.mesh, con.Clusters)
+			err := stream.Send(response)
+			if err != nil {
+				return err
+			}
+
+			if edsDebug {
+				log.Infof("EDS RES for %q clusters %v, Response: \n%s %s \n", peerAddr,
+					con.Clusters, response.String())
+			}
+		} else {
+			if edsDebug {
+				log.Infof("EDS empty clusters %v \n", peerAddr)
+			}
+		}
+	}
+}
+
+// EdsPushAll implements old style invalidation, generated when any rule or endpoint changes.
+// Primary code path is from v1 discoveryService.clearCache(), which is added as a handler
+// to the model ConfigStorageCache and Controller.
+func EdsPushAll() {
+	if edsDebug {
+		log.Infoa("EDS cache reset")
+	}
+	edsClusterMutex.Lock()
+	// Create a temp map to avoid locking the add/remove
+	tmpMap := map[string]*EdsCluster{}
+	for k, v := range edsClusters {
+		tmpMap[k] = v
+	}
+	version++
+	edsClusterMutex.Unlock()
+
+	for clusterName, edsCluster := range tmpMap {
+		updateCluster(clusterName, edsCluster)
+		for _, edsCon := range edsCluster.EdsClients {
+			edsCon.pushChannel <- true
```

Comment:
Because a EdsClient has multiple clusters, the double loops (line 373 and line 375) will cause multiple signals sent to the same edsCon if I understand it correctly.

Reply:
What is a cluster in this context? is the edsCluster corresponding to an envoy cluster? Line 167 and onwards suggests that's the case. 
I thought that an EdsClient serves EDS to an Envoy sidecar. In that case, there will be multiple clusters that needs to be configured. I might be wrong if the 'cluster' in here means differently.

Reply:
edsCluster matches an envoy cluster ( that at least one envoy has asked for ).

EdsClient is one envoy client (sidecar). Each cluster is watched by multiple sidecars. 

When a cluster is updated (when we have the event-driven stuff) - we'll only push to the clients watching that cluster.